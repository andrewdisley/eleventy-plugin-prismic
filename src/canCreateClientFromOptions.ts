import {
	PrismicPluginOptions,
	PrismicPluginOptionsWithClient,
	PrismicPluginOptionsWithEndpoint,
} from "./types";

/**
 * Checks if options are capable of creating a client
 *
 * @param options - Plugin options
 *
 * @returns `true` when options are capable of creating a client, `false` otherwise
 *
 * @internal
 */
export const canCreateClientFromOptions = (
	options: PrismicPluginOptions,
): options is
	| PrismicPluginOptionsWithClient
	| PrismicPluginOptionsWithEndpoint => {
	return (
		("endpoint" in options && typeof options.endpoint === "string") ||
		("client" in options && typeof options.client === "object")
	);
};
