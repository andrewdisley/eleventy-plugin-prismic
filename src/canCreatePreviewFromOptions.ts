import { PrismicPluginOptions, PrismicPluginOptionsWithPreview } from "./types";

/**
 * Checks if options are capable of creating a preview with 11ty Serverless
 *
 * @param options - Plugin options
 *
 * @returns `true` when options are capable of creating a preview with 11ty Serverless, `false` otherwise
 *
 * @internal
 */
export const canCreatePreviewFromOptions = (
	options: PrismicPluginOptions,
): options is PrismicPluginOptionsWithPreview => {
	return "preview" in options;
};
