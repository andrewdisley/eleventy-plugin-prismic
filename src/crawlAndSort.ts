import { Client } from "@prismicio/client";
import { asLink } from "@prismicio/helpers";
import { PrismicDocument } from "@prismicio/types";

import {
	PrismicPluginOptionsWithClient,
	PrismicPluginOptionsWithEndpoint,
} from "./types";
import { dPrismicClient } from "./lib/debug";

// Helper types
type SimpleDocuments = PrismicDocument | PrismicDocument[];
type I18nDocuments =
	| ({ __all: PrismicDocument[] } & Record<string, PrismicDocument>)
	| ({ __all: PrismicDocument[] } & Record<string, PrismicDocument[]>);

/**
 * Get all documents from given client and sort them
 *
 * @param client - A Prismic client instance
 * @param options - Options of the plugin
 *
 * @returns A promise resolving to a map of sorted documents
 *
 * @internal
 */
export const crawlAndSort = async (
	client: Client,
	options: PrismicPluginOptionsWithClient | PrismicPluginOptionsWithEndpoint,
): Promise<Record<string, SimpleDocuments> | Record<string, I18nDocuments>> => {
	const docs = await client.dangerouslyGetAll({
		// If the website is i18n, fetch for all languages
		lang: options.i18n ? "*" : undefined,
	});

	const fakeSingletons: string[] = [];
	const sortedDocs = docs.reduce(
		(
			collections:
				| Record<string, SimpleDocuments>
				| Record<string, I18nDocuments>,
			current,
		) => {
			// Resolve document URL if using a link resolver
			if (
				"linkResolver" in options &&
				typeof options.linkResolver === "function" &&
				!("url" in current && current.url)
			) {
				current.url = asLink(current, options.linkResolver);
			}

			let key: string | null = null;

			// If we're on a i18n site...
			if (options.i18n) {
				// ...set key to the language of the document
				key =
					typeof options.i18n === "object" && current.lang in options.i18n
						? options.i18n[current.lang]
						: current.lang;

				// If the collection is not yet defined
				if (!collections[current.type]) {
					collections[current.type] = {
						__all: [],
					};
				}
			}

			// Get current collection
			const get = () =>
				key
					? (collections[current.type] as I18nDocuments)[key]
					: collections[current.type];
			// Set current collection
			const set = (value: PrismicDocument | PrismicDocument[]) => {
				if (key) {
					(collections[current.type] as I18nDocuments)[key] = value;
					(collections[current.type] as I18nDocuments).__all.push(
						...(Array.isArray(value) ? value : [value]),
					);
				} else {
					collections[current.type] = value;
				}
			};
			// Push to current collection
			const push = (value: PrismicDocument) => {
				(get() as PrismicDocument[]).push(value);
				if (key) {
					(collections[current.type] as I18nDocuments).__all.push(value);
				}
			};

			if (
				"singletons" in options &&
				Array.isArray(options.singletons) &&
				options.singletons.includes(current.type) &&
				!fakeSingletons.includes(current.type) &&
				!get()
			) {
				set(current);
			} else {
				if (!get()) {
					set([]);
				} else if (!Array.isArray(get())) {
					dPrismicClient(
						"Document type %o was declared as a singleton but is not, converting to array gracefully",
						current.type,
					);

					fakeSingletons.push(current.type);

					if (key) {
						Object.entries(collections[current.type] as I18nDocuments)
							.filter(([key]) => key !== "__all")
							.forEach(([locale, document]) => {
								(collections[current.type] as I18nDocuments)[locale] = [
									document as PrismicDocument,
								];
							});
					} else {
						collections[current.type] = [get() as PrismicDocument];
					}
				}

				push(current);
			}

			return collections;
		},
		{},
	);

	// Give a pretty console overview of available documents
	const longestType =
		Math.max(...Object.keys(sortedDocs).map((type) => type.length), 8) + 2;
	dPrismicClient(
		[
			"Documents sorted, available types:",
			`  type${" ".repeat(longestType - 4)}   singleton   entries`,
			Object.entries(sortedDocs)
				.map(([type, documentsOrDocumentMap]) => {
					// Type
					let line = `  %o${" ".repeat(longestType - type.length)}`;

					// Singleton
					if (
						!(
							(options.i18n &&
								Object.values(documentsOrDocumentMap).every((v) =>
									Array.isArray(v),
								)) ||
							Array.isArray(documentsOrDocumentMap)
						)
					) {
						line = `${line} *${" ".repeat("singleton".length + 1)}`;
					} else {
						line = `${line} ${" ".repeat("singleton".length + 2)}`;
					}

					// Entries
					if (options.i18n) {
						line = `${line} ${Object.entries(documentsOrDocumentMap)
							.filter(([key]) => key !== "__all")
							.sort((a, b) => (a[0] > b[0] ? 1 : -1))
							.map(
								([locale, documents]) =>
									`${locale}:${
										Array.isArray(documents) ? documents.length : 1
									}`,
							)
							.join(", ")}`;
					} else {
						line = `${line} ${
							Array.isArray(documentsOrDocumentMap)
								? documentsOrDocumentMap.length
								: 1
						}`;
					}

					return line;
				})
				.join("\n"),
			"",
		].join("\n\n"),
		...Object.keys(sortedDocs),
	);

	return sortedDocs;
};
