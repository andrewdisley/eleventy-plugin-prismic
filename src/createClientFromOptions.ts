import nodeFetch from "node-fetch";
import { Client, createClient, getEndpoint } from "@prismicio/client";

import {
	PrismicPluginOptionsWithClient,
	PrismicPluginOptionsWithEndpoint,
} from "./types";

export const createClientFromOptions = (
	options: PrismicPluginOptionsWithClient | PrismicPluginOptionsWithEndpoint,
): Client => {
	return "client" in options
		? options.client
		: createClient(
				/** @see Regex101 expression: {@link https://regex101.com/r/GT2cl7/1} */
				/^(https?:)?\/\//gim.test(options.endpoint)
					? options.endpoint
					: getEndpoint(options.endpoint),
				{
					fetch: nodeFetch,
					...options.clientConfig,
				},
		  );
};
