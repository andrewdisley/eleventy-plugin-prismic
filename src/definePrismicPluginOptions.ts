import { PrismicPluginOptions } from "./types";

/**
 * Sugar function to get intellisense while writing configuration of the plugin
 *
 * @param options - The options to configure the plugin
 *
 * @returns the same options object
 *
 * @example
 * Basic usage
 *
 * ```
 * const prismicPluginOptions = definePrismicPluginOptions({
 *   endpoint: "my-repo",
 * });
 *
 * eleventyConfig.addPlugin(pluginPrismic, prismicPluginOptions);
 * ```
 */
export const definePrismicPluginOptions = (
	options: PrismicPluginOptions,
): PrismicPluginOptions => options;
