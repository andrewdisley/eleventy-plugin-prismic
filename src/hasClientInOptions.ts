import {
	PrismicPluginOptions,
	PrismicPluginOptionsWithClient,
	PrismicPluginOptionsWithEndpoint,
} from "./types";

export const hasClientInOptions = (
	options: PrismicPluginOptions,
): options is
	| PrismicPluginOptionsWithClient
	| PrismicPluginOptionsWithEndpoint => {
	return (
		("endpoint" in options && typeof options.endpoint === "string") ||
		("client" in options && typeof options.client === "object")
	);
};
