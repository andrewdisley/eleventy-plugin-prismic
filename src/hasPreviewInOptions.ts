import { PrismicPluginOptions, PrismicPluginOptionsWithPreview } from "./types";

export const hasPreviewInOptions = (
	options: PrismicPluginOptions,
): options is PrismicPluginOptionsWithPreview => {
	return "preview" in options;
};
