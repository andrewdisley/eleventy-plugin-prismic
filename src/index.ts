export { pluginPrismic } from "./pluginPrismic";
export { definePrismicPluginOptions } from "./definePrismicPluginOptions";

export { crawlAndSort } from "./crawlAndSort";

export {
	asText,
	asHTML,
	asLink,
	asDate,
	image,
	embed,
	injectShortcodes,
} from "./shortcodes";

export { link, injectPairedShortcodes } from "./pairedShortcodes";

export type { PrismicPluginOptions } from "./types";
