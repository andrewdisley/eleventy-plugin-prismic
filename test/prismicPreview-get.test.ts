import { it, expect, beforeAll, afterAll } from "vitest";
import * as mswNode from "msw/node";

import { createMockQueryHandler } from "./__testutils__/createMockQueryHandler";
import { createMockRepositoryHandler } from "./__testutils__/createMockRepositoryHandler";
import { createDocument } from "./__testutils__/createDocument";

import { prismicPreview } from "../src";

const options = {
	endpoint: "prismicpreview-get-test-ts",
	preview: {
		name: "preview",
		inputDir: "./test/__fixtures__/preview",
		functionsDir: "./test/__fixtures__/preview/functions",
	},
};

const headers = {
	cookie: `${options.endpoint}.prismic.io=${encodeURIComponent(
		JSON.stringify({ preview: "foo" }),
	)}`,
};

const server = mswNode.setupServer(
	createMockRepositoryHandler(options.endpoint),
	createMockQueryHandler(options.endpoint, [
		createDocument({ lang: "en-us", url: "/foo" }),
	]),
);
beforeAll(() => server.listen({ onUnhandledRequest: "error" }));
afterAll(() => server.close());

it("returns previewed page when in a preview session", async () => {
	expect(
		await prismicPreview.get("/preview/foo", {}, headers, options),
	).toStrictEqual({
		statusCode: 200,
		body: `<h1>foo</h1>
`,
		headers: {
			"Content-Type": "text/html; charset=UTF-8",
		},
	});
});

it("returns fallback page when not in a preview session", async () => {
	const response = await prismicPreview.get("/preview/404", {}, {}, options);

	expect(response).toMatchSnapshot();
	expect(response.statusCode).toBe(404);
});

it("returns fallback page when previewed page is not found", async () => {
	const response = await prismicPreview.get(
		"/preview/404",
		{},
		headers,
		options,
	);

	expect(response).toMatchSnapshot();
	expect(response.statusCode).toBe(404);
});
