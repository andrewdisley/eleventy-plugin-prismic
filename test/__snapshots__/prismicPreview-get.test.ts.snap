// Vitest Snapshot v1

exports[`returns fallback page when not in a preview session 1`] = `
{
  "body": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"UTF-8\\">
<meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
<title>Not found</title>
<style>
	body, h1 {
		font-family: monospace;
		font-weight: 400;
		font-size: 16px;
	}
	ul {
		list-style-type: decimal;
	}
</style>
</head>
<body>
<section>
	<h1>404 - Not Found.</h1>
	<p>This probably means one of the following:</p>
	<ul>
		<li>You are entering a preview session and it is still loading. Hang on!</li>
		<li>You are not in an ongoing preview session or just exited one.</li>
	</ul>
	<p><a href=\\"/\\">Get back to home</a></p>
</section>
<script async defer src=\\"https://static.cdn.prismic.io/prismic.js?new=true&repo=prismicpreview-get-test-ts.prismic.io\\"></script>
</body>
</html>
",
  "headers": {
    "Content-Type": "text/html; charset=UTF-8",
  },
  "statusCode": 404,
}
`;

exports[`returns fallback page when previewed page is not found 1`] = `
{
  "body": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"UTF-8\\">
<meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
<title>Not found</title>
<style>
	body, h1 {
		font-family: monospace;
		font-weight: 400;
		font-size: 16px;
	}
	ul {
		list-style-type: decimal;
	}
</style>
</head>
<body>
<section>
	<h1>404 - Not Found.</h1>
	<p>This probably means one of the following:</p>
	<ul>
		<li>You are entering a preview session and it is still loading. Hang on!</li>
		<li>You are not in an ongoing preview session or just exited one.</li>
	</ul>
	<p><a href=\\"/\\">Get back to home</a></p>
</section>
<script async defer src=\\"https://static.cdn.prismic.io/prismic.js?new=true&repo=prismicpreview-get-test-ts.prismic.io\\"></script>
</body>
</html>
",
  "headers": {
    "Content-Type": "text/html; charset=UTF-8",
  },
  "statusCode": 404,
}
`;
